### PR TITLE
Teiid class to connect and query the teiid db

### PR DESCRIPTION
**Pull request by yashvardhannanavati**

### CODE REVIEW

**Comment by yashvardhannanavati** [comment]
@mprahl Can you please review this?

File: purview/utils/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from purview.utils.teiid import Teiid
+
+__all__ = ['Teiid']
```
Comment (mprahl) [inline purview/utils/__init__.py]: Can you please remove this? We'll hopefully never do `from utils import *` and it's a pain to maintain.

File: purview/utils/teiid.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+# Taken from MARS project
+import logging
+
+import psycopg2
+
+LOG = logging.getLogger(__name__)
```
Comment (mprahl) [inline purview/utils/teiid.py]: Can we use a global `log` variable defined in `purview/__init__.py`? This way we don't configure logging per file.

File: purview/utils/teiid.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# Taken from MARS project
```
Comment (mprahl) [inline purview/utils/teiid.py]: You can change this to `# Inspired by the MARS project`.

File: purview/utils/teiid.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self, host, port, username, password):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.password = password
+
+        # a dict mapping db names to cursors
+        self._connections = {}
+
+    def get_connection(self, dbname):
```
Comment (mprahl) [inline purview/utils/teiid.py]: Can you rename `dbname` to `db_name`? That is a convention in Python code.

File: purview/utils/teiid.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        """
```
Comment (mprahl) [inline purview/utils/teiid.py]: Can you change this to the standard format for doc strings?
```python
"""
Return an existing psycopg2 connection and establish it if needed
:arg db_name: a string of the database name to get a connection to
:return: a psycopg2 connection
"""
```

File: purview/utils/teiid.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+
+class Teiid(object):
+
+    def __init__(self, host, port, username, password):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.password = password
+
```
Comment (mprahl) [inline purview/utils/teiid.py]: Bike-shedding: Can you remove this new line?

File: purview/utils/teiid.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        Args:
+            dbname: database name for which to get connection
+
+        Returns: psycopg2 connection
+        """
+        if dbname in self._connections:
+            return self._connections[dbname]
+
+        LOG.info("Connecting to Teiid host %s:%s", self.host, self.port)
```
Comment (mprahl) [inline purview/utils/teiid.py]: Can you make this a debug log? This seems very verbose for an info log.

File: purview/utils/teiid.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            database=dbname,
+            host=self.host,
+            port=str(self.port),
+            user=self.username,
+            password=self.password,
+        )
+
+        # Teiid does not support setting this value at all and unless we
+        # specify ISOLATION_LEVEL_AUTOCOMMIT (zero), psycopg2 will send a
+        # SET command the teiid server doesn't understand.
```
Comment (mprahl) [inline purview/utils/teiid.py]: "the teiid" => "to the Teiid"

File: purview/utils/teiid.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        conn.set_isolation_level(0)
+
+        self._connections[dbname] = conn
+        return conn
+
+    def query(self, sql, db='public', retry=1):
+        """Send the SQL query to Teiid and return the rows
```
Comment (mprahl) [inline purview/utils/teiid.py]: Can you change this to the standard format for doc strings?
```python
"""
Send the SQL query to Teiid and return the rows as a list
:param sql: a string of the SQL query to send to the database
:kwarg db: the database name to query on
:kwarg retry: an int of the  number of times to retry a failed query
:return: a list of rows from Teiid. Each row is a dictionary
    with the column header as the key.
"""
```

File: purview/utils/teiid.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                break
+            except psycopg2.extensions.QueryCanceledError as err:
+                LOG.warning("Teiid query failed")
+        else:
+            raise err
+
+        data = cursor.fetchall()
+        cols = [t[0] for t in cursor.description or []]  # column header names
+        LOG.debug("Found the following columns: %s", cols)
+        LOG.info("Received %s rows from Teiid", len(data))
```
Comment (mprahl) [inline purview/utils/teiid.py]: This should be debug information as well.

File: purview/utils/teiid.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
```
Comment (mprahl) [inline purview/utils/teiid.py]: Can you remove this new line?

File: purview/utils/teiid.py (hunk 2 above)
Comment (mikebonnet) [reply purview/utils/teiid.py]: Isn't logging-per-file preferable? It gives you a better indication of where the log message is coming from, and finer-grained control over logging levels.

**Comment by yashvardhannanavati** [comment]
@mprahl  I made the changes, can you please check and merge this?

**Comment by yashvardhannanavati** [comment]
FYI, I'll submit a separate PR for the log variable

File: purview/utils/teiid.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            except psycopg2.extensions.QueryCanceledError as err:
+                log.warning("Teiid query failed")
+        else:
+            raise err
+
+        data = cursor.fetchall()
+        cols = [t[0] for t in cursor.description or []]  # column header names
+        log.debug("Found the following columns: {}".format(cols))
+        log.debug("Received {} rows from Teiid".format(len(data)))
+        # bulid a return array with all columns
```
Comment (mprahl) [inline purview/utils/teiid.py]: "bulid" => "build"

**Comment by yashvardhannanavati** [comment]
I think my changes made @mikebonnet 's comment disappear. Sorry about that. 
Here it is "Isn't logging-per-file preferable? It gives you a better indication of where the log message is coming from, and finer-grained control over logging levels. "

**Comment by mprahl** [comment]
@yashvardhannanavati looks great. Just reimplement the logging as @mikebonnet asked for and fix the typo in the comment.

**Pull request merged**